sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -133,8 +135,21 @@ abstract class AbstractGpuJoinIterator(
   private def nextCbFromGatherer(): Option[ColumnarBatch] = {
     withResource(new NvtxWithMetrics(gatherNvtxName, NvtxColor.DARK_GREEN, joinTime)) { _ =>
       val ret = gathererStore.map { gather =>
-        val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
-        gather.gatherNext(nextRows)
+        gather.checkpoint
+        gather.allowSpilling()
+        withRetryNoSplit[ColumnarBatch] {
+          try {
+            val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
+            gather.gatherNext(nextRows)
+          } catch {
+            case retryOOM: RetryOOM =>
```

Comment:
Should we try to make this a part of withRetryNoSplit?  Like an onRetry callback that we can pass in? Perhaps something like the following.

```
  def withRetryNoSplit[K](fn: => K)(onRetry: => Unit = noopOnRetry): K = {
```

Or perhaps a `checkpointable` trait so instead of closing the input we checkpoint and restore it.  The reason I mention this is because we want to do something similar for GpuRandom in the future.

Reply:
Thanks @revans2!  @abellina and I talked about this last week as well and I agree it would be a good improvement.  I will file a follow-on PR to add this.

Reply:
I filed issue #7943 for this.